Styling Next.js with Styled JSX (#285)
zeit/next-site: www/pages/blog/styling-next-with-styled-jsx.mdx

function Alert(props) {
  return (
    <div className="alert">
      {props.children}
      <style jsx>{`
        .alert {
          display: inline-block;
          padding: 20px;
          border-radius: 8px;
          background: ${props.type == 'warning' ? '#fff3cd' : '#eee'};
        }
      `}</style>
    </div>
  )
}

export default Alert
```

If the `Alert` component is passed a `type` prop with a `warning` value like `<Alert type="warning">This is an important message</Alert>`, the component will have an orange background. Without specifying the type prop, the background would fallback to the default grey color.

Alternative approach to adapting styles based on props is to apply different class names based on the prop value:

```js
// components/Alert.js
function Alert(props) {
  return (
    <div className={props.type == 'warning' ? 'alert warning' : 'alert'}>
      {props.children}
      <style jsx>{`
        .alert {
          display: inline-block;
          padding: 20px;
          border-radius: 8px;
          background: #eee;
        }
        .alert.warning {
          background: #fff3cd;
        }
      `}</style>
    </div>
  )
}

export default Alert
```

> ## Using Sass, Less or PostCSS with Styled JSX

Reviewer: Let's leave this part out and focus on the core of styled-jsx features 👌

import withPost from '../../components/blog/with-post'
import Image from '../../components/image'
import { meta } from '../../blog/styling-next-with-styled-jsx-preview.mdx'

export default withPost({ ...meta })

[Styled JSX](https://github.com/zeit/styled-jsx) is a CSS-in-JS library that allows you to write isolated scoped CSS to style your components. This blog post will help you get started with using Styled JSX in Next.js.

> Next.js includes Styled JSX by default, so getting started is as simple as adding a <br /> `<style jsx>` tag into your existing Next.js component and writing CSS inside of it:

Reviewer 2: I'd add a title before this: `Installation`

Just by adding a single `jsx` attribute to a `<style>` element you can write standard CSS that gets automatically scoped to the component and auto prefixed. `<style jsx>` elements should be placed inside the root element of the component.

## Adding Global styles

Most projects need some global styles to style the body element or provide css resets. Styled JSX allows us to add global styles using `<style jsx global>`. For example:

```js
// pages/index.js
function Home() {
  return (
    <div className="container">
      <h1>Hello Next.js</h1>
      <p>Let's explore different ways to style Next.js apps</p>
      <style jsx>{`
        .container {
          margin: 50px;
        }
        p {
          color: blue;
        }
      `}</style>
      <style jsx global>{`
        p {
          font-size: 20px;
        }
      `}</style>
    </div>
  )
}

export default Home
```

> This will apply 20px font-size to all `<p>` tags in this specific page.

Reviewer 2: ```suggestion
This will apply a 20px font-size to all `<p>` tags in this specific page.
```

> To add global styles that will be applied to all pages of our app, a good approach is to create a layout component, add global styles there and wrap all pages with it.

Reviewer 2: ```suggestion
To apply global styles to all pages in our app, a good approach is to first create a layout component with the global styles, and then to wrap all other pages with it.
```

Using a layout component provides the flexibility to apply a certain set of styles to some pages while still allowing a different style for others:

```js
// components/Layout.js
function Layout(props) {
  return (
    <div className="page-layout">
      {props.children}
      <style jsx global>{`
        body {
          margin: 0;
          padding: 0;
          font-size: 18px;
          font-weight: 400;
          line-height: 1.8;
          color: #333;
          font-family: sans-serif;
        }
        h1 {
          font-weight: 700;
        }
        p {
          margin-bottom: 10px;
        }
      `}</style>
    </div>
  )
}

export default Layout
```

> In Next.js you can load the layout once for all of your pages by creating a custom [`App` component](https://nextjs.org/docs/#custom-app) in `pages/_app.js`, importing the `Layout` component and adding it to the render method like this:

Reviewer 2: ```suggestion
In Next.js, you can load the layout at once for all of your pages by creating a custom [`App` component](https://nextjs.org/docs/#custom-app) within `pages/_app.js`, importing the `Layout` component, and then adding it to the render method as follows:
```

```js
// pages/_app.js
import React from 'react'
import App, { Container } from 'next/app'
import Layout from '../components/Layout'

class MyApp extends App {
  render() {
    const { Component, pageProps } = this.props

    return (
      <Container>
        <Layout>
          <Component {...pageProps} />
        </Layout>
      </Container>
    )
  }
}

export default MyApp
```

### Writing styles in external files

> We can also place css styles in external files outside of the component.

Reviewer 2: ```suggestion
We can also place styles in external files outside of the component.
```

> For example, we can move our global styles from the `Layout` component into a separate file like this:

Reviewer 2: ```suggestion
For example, we can move our global styles from the `Layout` component into a separate file as follows:
```

```js
// styles/global.js
import css from 'styled-jsx/css'

export default css`
  body {
    margin: 0;
    padding: 0;
    font-size: 18px;
    font-weight: 400;
    line-height: 1.8;
    color: #333;
    font-family: sans-serif;
  }
  h1 {
    font-weight: 700;
  }
  p {
    margin-bottom: 10px;
  }
`
```

and then include the styles back into the `Layout` component by importing the file with styles:

```js
// components/Layout.js
import globalStyles from '../styles/global.js'

function Layout(props) {
  return (
    <div className="page-layout">
      {props.children}
      <style jsx global>
        {globalStyles}
      </style>
    </div>
  )
}

export default Layout
```

### One-off global selectors

> The styles that we add to a component using `<style jsx>` affect only the elements inside that component, they won't affect child components.

Reviewer 2: ```suggestion
The styles that we add to a component using `<style jsx>` affect only the elements inside that component. They don't affect child components.
```

> Sometimes we may need to override a certain style of a child component. To be able to do this, Styled JSX provides `:global()` to get _one-off global selectors_.

Reviewer 2: ```suggestion
At times, we may need to override a certain style of a child component. To be able to do this, Styled JSX provides `:global()` to get _one-off global selectors_.
```

For example, let's say we have a `<Widget>` component that contains a button with a class name `btn` and we want to change the colors of this button only when the widget is imported on the homepage. We can do it like this:

```js
// pages/index.js
import Widget from '../components/Widget'

function Home() {
  return (
    <div className="container">
      <h1>Hello Next.js</h1>
      <Widget />
      <style jsx>{`
        .container {
          margin: 50px;
        }
        .container :global(.btn) {
          background: #000;
          color: #fff;
        }
      `}</style>
    </div>
  )
}

export default Home
```

## Dynamic styles

> Being able to adapt the styling of a component based on its props is one of the big advantages of `CSS-in-JS` libraries compared to other styling solutions.

Reviewer 2: ```suggestion
Compared to other solutions, being able to adapt the styling of a component based on its props is one big advantage of `CSS-in-JS` libraries.
```

With Styled JSX we can do it like this:

```js
// components/Alert.js
function Alert(props) {
  return (
    <div className="alert">
      {props.children}
      <style jsx>{`
        .alert {
          display: inline-block;
          padding: 20px;
          border-radius: 8px;
        }
      `}</style>
      <style jsx>{`
        .alert {
          background: ${props.type == 'warning' ? '#fff3cd' : '#eee'};
        }
      `}</style>
    </div>
  )
}

export default Alert
```

If the `Alert` component is passed a `type` prop with a `warning` value like `<Alert type="warning">This is an important message</Alert>`, the component will have an orange background. Without specifying the type prop, the background would fallback to the default grey color.

Note that we placed the dynamic style into a separate `<style jsx>` tag. This isn't required, but it's recommended to split up static and dynamic styles so that only the dynamic parts are recomputed.

> Alternative approach to adapting styles based on props is to apply different class names based on the prop value:

Reviewer 2: ```suggestion
An alternate approach to adapting styles based on props is to apply different class names based on the prop value as shown below:
```

```
{
  "presets": [
    [
      "next/babel",
      {
        "styled-jsx": {
          "plugins": ["styled-jsx-plugin-sass"]
        }
      }
    ]
  ]
}
```

Now we can use Sass features in `<style jsx>` tags in our Next.js components. For example:

```js
import Link from 'next/link'

function Menu(props) {
  return (
    <ul className="menu">
      <li>
        <Link href="/">
          <a>Home</a>
        </Link>
      </li>
      <li>
        <Link href="/about">
          <a>About</a>
        </Link>
      </li>
      <style jsx>{`
        $displayType: ${props.horizontal ? 'inline-block' : 'block'};
        ul {
          margin: 20px 0;
          padding: 0;
          li {
            list-style: none;
            display: $displayType;
            a {
              padding: 10px;
            }
          }
        }
      `}</style>
    </ul>
  )
}

export default Menu
```

Here we are using nested styles to style the menu list and a Sass variable that changes based on the prop passed to the `Menu` component.

The same approach can be used to implement Less and other CSS preprocessors.

> Styled JSX provides some additional features that we didn't cover here. To learn more, check its [GitHub repository](https://github.com/zeit/styled-jsx).

Reviewer 2: ```suggestion
In this blog post, we covered how one get started with Styled JSX. To learn more about additional features, be sure to [check it out on GitHub](https://github.com/zeit/styled-jsx). 
```
